### ISSUE TITLE
Create an es2015 ready build

### ISSUE THREAD

**Issue by hourliert** [inversify/InversifyJS#138]
Hi,
It would be nice if this library could be exported as an es2015 library. (In that way, it could work out of the box with `webpack2` or `rollup`).
## Expected Behavior
- Keep the `dist` folder containing `inversify.js`.
- Create new folders (`es` or `lib` for instance) including transpiled typescript files without concatenating them.
## Possible Solution

We could take inspiration on [redux](https://github.com/reactjs/redux) which transpile ES2015 to ES5 via babel in 2 separates folders `lib` and `es`. 
I have also created a project [builder-react-comp](https://github.com/hourliert/builder-react-comp) which transpile my TS react component into 2 versions. One compatible with webpack2 and rollup and one compatible with webpack1 or browserify.
An use case of this project is available [here](https://github.com/hourliert/react-component-seed).
## Context

The main issue if we don't do that is that when you are building a project that has a dependency on inversify with `webpack1`. You get this compilation warning:

```
WARNING in ./~/inversify/dist/inversify.js
Critical dependencies:
7:482-489 This seems to be a pre-built javascript file. Though this is possible, it's not recommended. Try to require the original source to get better results.
 @ ./~/inversify/dist/inversify.js 7:482-489
```

Thanks in advance.

**Comment by remojansen** [comment]
Thanks a lot for reporting this issue. I have one question. I like the Redux style:
- dist (bundled + sourcemaps and compressed bundled + sourcemaps)
- lib (es5 not compressed)
- es (es6 not compressed)

What about sourcemaps for lib and es6? Should they be included? I took a look top redux and it doesn't include sourcemaps (even for the compressed file in dist).

Thanks!

**Comment by hourliert** [comment]
Hi,

You're right. I have never noticed that (maybe because I never had to debug something related to redux).
I guess, we still could include source maps like in the dist folder as long as they are not inline.

We could use the typescript compiler directly (without using gulp) to output these 2 folders. Or just create new gulp tasks to be standardized.

What do you think? I could do the PR if you want.

**Comment by remojansen** [comment]
I think is good idea two add them as a gulp task. I was actually thinking about adding an extra compilation targeting es6 a couple of days ago. 

I wasn't thinking about including the output to the npm releases, I was thinking about catching potential duplicated definitions errors caused by the `es6.d.ts` file that is included by the compiler when we target es6.

We are including sourcemaps already in the dist folder using `.map` files. So we just need:
- [ ] Two more tasks to generate the `lib` and `es` folders. I haven't test it but should be similar to the following:

```
var tsLibProject = tsc.createProject("tsconfig.json", { target : "ES5" });

gulp.task("build-lib", function() {
    return gulp.src([
        "src/**/*.ts"
    ])
    .pipe(tsc(tsLibProject ))
    .on("error", function (err) {
        process.exit(1);
    })
    .js.pipe(gulp.dest("lib/"));
});

var tsEsProject = tsc.createProject("tsconfig.json", { target : "ES6" });

gulp.task("build-es", function() {
    return gulp.src([
        "src/**/*.ts"
    ])
    .pipe(tsc(tsEsProject))
    .on("error", function (err) {
        process.exit(1);
    })
    .js.pipe(gulp.dest("es/"));
});
```
- [ ] Add this two new tasks to [the `build` task](https://github.com/inversify/InversifyJS/blob/master/gulpfile.js#L181-L189) so it runs on CI.
- [ ] Add the `lib` and `es` folders to the `.gitignore` file.

PRs are more than welcome, I think I haven't left anything behind but please add a comment if you are unsure about anything.

**Comment by hourliert** [comment]
Seems good. The only remaining thing is how should we modify `package.json`?

In redux, they are a `main` key pointing to `lib/index.js` and a `jsnext:main` key pointing to `es/index.js` [here](https://github.com/reactjs/redux/blob/6317727a32b7e45bbfc0ca2f48d4bab3137f5892/package.json#L5L6).

The file `dist/redux.js` is not included in `package.json`.

For inversify, what about `dist/inversify.js`?

**Comment by remojansen** [comment]
OK this makes sense now. They don't have sourcemaps because they never compress and they don't have `dist` folder either. Let's leave the dist `folder` there for now. Is handy if someone wants to test it without a build script.

As you say we will need to update the `package.json` as you have said:

```
"main": "lib/inversify.js",
"jsnext:main": "es/inversify.js",
```

I wasn't aware of the `jsnext` field is nice to learn about it :smile:

**Comment by ulrichb** [comment]
In the just released `2.0.0-rc.14`, the `dist`-dir disappeared. 

Is this as intended?

**Comment by remojansen** [comment]
As I have explained at the chat:

> We have removed it because most people is using a build tool like webpack and browserify
> the `dist` folder was good for being consumed directly in browsers but it was causing some problems that was preventing us from being able to auto-generate the dts files and simplify the installation process.
